Iss23

ericbarefoot opened the pull request:
This change solves a problem with logging from the onboard ATC. The onboard ATC allows readings for analog voltage up to 18 bits of precision, but the typical `Margay::run()` method cannot access values from this device. 

This is because the ATC actually communicates with I2C, and so external sensors (which are typically read from I2C) might have address clashes. `Margay::run()` solves this by turning the external I2C bus on, reading data, then turning it off, but we want to also read from the internal I2C in the same breath! 

Thus, we have now added a switch to the `Margay::getVoltage()` method to test if the external I2C bus is on, and if it is, turn it off, take a reading from the ADC, then turn external I2C back on again. 

Should close #23

awickert reviewed (changes requested):
Hey @ericbarefoot,

Thanks for all this! The `initADC()` function looks like a good addition.

Now for a few thoughts and requested changes, and I imagine that @bschulz1701 may have some as well.

1. I can see your idea behind `bool extI2COn = digitalRead(SD_CD);` being that you can first check whether internal or external I2C communcations are shunted to be open. However, the SD card communicates over SPI, and so this probably won't work.
2. Related to (1), could you add additional comments about why you're doing what you're doing in the code? This would be helpful: I think I just knew what was going on re:(1) because you told me about your idea, which in principle, is good.
3. In order to execute your idea in (1), we could establish a variable with global scope that stores this I2C internal/external state, and then switch this inside the `externalI2C()` function. (Or we could change the `externalI2C` function to a `switchI2C` function, with `EXTERNAL` or `INTERNAL` being a defined boolean set of parameters.
4. For general tidiness, I would include `return val;` outside of the `if`-statement. Indeed, I might refactor the whole thing to have `if`s wrapping the on/off toggles and the same code within, for closer readability to the true logical flow.

ericbarefoot commented:
ok @awickert, this should do what you mentioned, I have fixed the typo, added some comments, and established a separate function to "close the farm gate," dependent on the state at the beginning of the function. This relies on a class-wide variable that I introduced that stores the state of the I2C bus

awickert commented on src/Margay.cpp:
```diff
@@ -715,22 +729,25 @@ void Margay::blinkGood()
+	switchExternalI2C(OFF);
+	initADC(18); // initialize ADC
+	val = adc.GetVoltage();
+	switchExternalI2C(ON); // re-allow external I2C		
```
Remove: No longer needed with the farmGateI2C function
```suggestion
```

awickert commented on src/Margay.cpp:
```diff
@@ -381,6 +384,9 @@ void Margay::I2Ctest()
 	if(I2C_Test) Serial.println("PASS");
+
+	// make sure I2C Bus is returned to initial state
+	closeFarmGateI2C(initialStateI2C);
```
Can open or close gate, so:
```suggestion
	farmGateI2C(initialStateExternI2C);
```

awickert commented on src/Margay.cpp:
```diff
@@ -715,22 +729,25 @@ void Margay::blinkGood()
+	switchExternalI2C(ON); // re-allow external I2C		
+
+	// make sure I2C Bus is returned to initial state
+	closeFarmGateI2C(initialStateI2C);
```
Can open or close gate, so:
```suggestion
	farmGateI2C(initialStateExternI2C);
```

awickert commented on src/Margay.cpp:
```diff
@@ -800,30 +817,60 @@ void Margay::resetWDT()  //Send a pulse to "feed" the watchdog timer
 	delay(1); // Any time needed to switch states; may not be necessary
 }
 
+void Margay::closeFarmGateI2C(bool initStateI2C)
```
Can open or close gate, so rename:
```suggestion
void Margay::farmGateI2C(bool initStateI2C)
```

awickert commented on src/Margay.cpp:
```diff
@@ -800,30 +817,60 @@ void Margay::resetWDT()  //Send a pulse to "feed" the watchdog timer
+	switchExternalI2C(OFF);
+
+	// make sure I2C Bus is returned to initial state
+	closeFarmGateI2C(initialStateI2C);
```
Can open or close gate, so rename:
```suggestion
	farmGateI2C(initialStateExternI2C);
```

awickert commented on src/Margay.cpp:
```diff
@@ -715,22 +729,25 @@ void Margay::blinkGood()
+	// bus communication by taking a voltage reading. So we have logic here to make the switch.
+
+	// first check whether external I2C connnections are on by testing pin I2C_SW (HIGH is on)
+	bool initialStateI2C = digitalRead(I2C_SW);
```
```suggestion
	bool initialStateExternalI2C = digitalRead(I2C_SW);
```

awickert commented on src/Margay.cpp:
```diff
@@ -715,22 +729,25 @@ void Margay::blinkGood()
+	// sensors, need access to the ADC. However, we do not want to change the state of the I2C 
+	// bus communication by taking a voltage reading. So we have logic here to make the switch.
+
+	// first check whether external I2C connnections are on by testing pin I2C_SW (HIGH is on)
```
```suggestion
	// first check whether external I2C connnections are on by testing pin I2C_SW (HIGH is on)
	// When the external I2C connection is on, the internal I2C connection is cut off.
```

awickert commented on src/Margay.cpp:
```diff
@@ -800,30 +817,60 @@ void Margay::resetWDT()  //Send a pulse to "feed" the watchdog timer
-	data = (*update)(); //Run external update function
-	externalI2C(OFF);
+	
+	bool initialStateI2C = digitalRead(I2C_SW);
```
```suggestion
	bool initialStateExternI2C = digitalRead(I2C_SW);
```

awickert commented on src/Margay.cpp:
```diff
@@ -800,30 +817,60 @@ void Margay::resetWDT()  //Send a pulse to "feed" the watchdog timer
+
+	switchExternalI2C(ON);
+	data = (*update)(); //Run external update function	
+	switchExternalI2C(OFF);
```
[redundant w/ farm gate]
```suggestion
```

awickert commented on src/Margay.cpp:
```diff
@@ -348,7 +348,10 @@ int Margay::begin(String header_)
 void Margay::I2Ctest()
 {
-	externalI2C(ON);
+	bool initialStateI2C = digitalRead(I2C_SW);
```
```suggestion
	bool initialStateExternI2C = digitalRead(I2C_SW);
```

awickert reviewed (changes requested):
Could you check these function renames for general code readability?
* `farmGate` function will both "open" and "close" it
* `initialState` is specifically of the *external* I2C, with the internal connection being toggled opposite
